packages/fetchai/connections/p2p_noise/noise_node.go

```diff
@@ -0,0 +1,188 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"os/signal"
+	//"strings"
+	"time"
+	"errors"
+	"github.com/perlin-network/noise"
+	"github.com/perlin-network/noise/kademlia"
+  	aea "./aea"
+)
+
+// check panics if err is not nil.
+func check(err error) {
+	if err != nil {
+		panic(err)
+	}
+}
+
+
+// An initial noise p2p node for AEA's fetchai/p2p-noise/0.1.0 connection
+func main() {
+	
+	// Create connection to aea
+	agent := aea.AeaApi{}
+	check(agent.Init())
+
+	// Create a new configured node.
+  	host, port := agent.Uri()
+  	key, err := noise.LoadKeysFromHex(agent.PrivateKey())
+	
+  	node, err := noise.NewNode(
+		noise.WithNodeBindHost(host),
+		noise.WithNodeBindPort(port),
+		noise.WithNodeAddress(""),
+    	noise.WithNodePrivateKey(key),
+	)
+	check(err)
+
+	// Release resources associated to node at the end of the program.
+	defer node.Close()
+
+	// Register Envelope message
+	node.RegisterMessage(aea.Envelope{}, aea.UnmarshalEnvelope)
+
+	// Register a message handler to the node.
+	node.Handle(func (ctx noise.HandlerContext) error {
+    	return handle(ctx, agent)
+ 	 })
+
+	// Instantiate Kademlia.
+	events := kademlia.Events{
+		OnPeerAdmitted: func(id noise.ID) {
+			fmt.Printf("[noise-p2p][info] Learned about a new peer %s(%s).\n", id.Address, id.ID.String())
+		},
+		OnPeerEvicted: func(id noise.ID) {
+			fmt.Printf("[noise-p2p][info] Forgotten a peer %s(%s).\n", id.Address, id.ID.String())
+		},
+	}
+
+	overlay := kademlia.New(kademlia.WithProtocolEvents(events))
+
+	// Bind Kademlia to the node.
+	node.Bind(overlay.Protocol())
+
+	// Have the node start listening for new peers.
+	check(node.Listen())
+
+	//
+	//fmt.Printf("[noise-p2p][info] started node %s (%s).\n", node.ID().Address, node.ID().ID.String())
+
+	// Ping entry node to initially bootstrap, if non genesis
+	if len(agent.EntryUris()) > 0 {
+		check(bootstrap(node, agent.EntryUris()...))
+	}
+
+	// Attempt to discover peers if we are bootstrapped to any nodes.
+	discover(overlay)
+
+	// Once overlay setup, connect to agent
+	check(agent.Connect())
+
+	// Receive envelopes from agent
```

Comment:
maybe slightly more explicit: `// Receive envelopes from agent and forward to peer`